snc-library.sh

```diff
@@ -26,6 +26,27 @@ function download_oc() {
     fi
 }
 
+function patch_operators_and_create_new_release() {
+    upstream_registry=quay.io/crcont
+    
+    kao_image=$(${OC} adm release info -a ${OPENSHIFT_PULL_SECRET_PATH} ${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE} --image-for=cluster-kube-apiserver-operator)
+    kao_vcs_ref=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${kao_image} -ojson | jq -r '.config.config.Labels."vcs-ref"')
+    # As of now `oc adm release new` not able to parse images which have multiple arch manifest file so we first need to get the digest of the image for specific
+    # arch and then use that digest with `oc adm release new`.
+    kao_image_diget=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${upstream_registry}/openshift-crc-cluster-kube-apiserver-operator:${kao_vcs_ref} --filter-by-os=linux/${yq_ARCH} -ojson | jq -r .digest)
+
+    kcmo_image=$(${OC} adm release info -a ${OPENSHIFT_PULL_SECRET_PATH} ${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE} --image-for=cluster-kube-controller-manager-operator)
+    kcmo_vcs_ref=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${kcmo_image} -ojson | jq -r '.config.config.Labels."vcs-ref"')
+    kcmo_image_diget=$(${OC} image info -a ${OPENSHIFT_PULL_SECRET_PATH} ${upstream_registry}/openshift-crc-cluster-kube-controller-manager-operator:${kcmo_vcs_ref} --filter-by-os=linux/${yq_ARCH} -ojson | jq -r .digest)
+
+    release_hash=$(${OC} adm release info -a ${OPENSHIFT_PULL_SECRET_PATH} --commits ${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE} -ojsonpath='{.digest}' |  cut -d":" -f2)
+    ${OC} adm release new -a ${OPENSHIFT_PULL_SECRET_PATH} --from-release=${OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE} \
+	    cluster-kube-apiserver-operator=${upstream_registry}/openshift-crc-cluster-kube-apiserver-operator@${kao_image_diget} \
+	    cluster-kube-controller-manager-operator=${upstream_registry}/openshift-crc-cluster-kube-controller-manager-operator@${kcmo_image_diget} \
+	    --to-image=quay.io/crcont/ocp-release:${release_hash}
```

Comment:
Since we have to generate the kao/kcmo images beforehand, I wonder if we should do this work in advance as well instead of redoing it for each `snc.sh` run?

Reply:
This should happen for arch specific so better to do it here (snc.sh) then beforehand.

Reply:
it's not possible to generate an aarch64 release image on an amd64 machine?

Reply:
`oc adm release new` doesn't have any arch or `filter-by-os` options as of now it happen in 2 steps, first get the digest and then we run the `oc adm release new` command.

Reply:
Do you mean that using internal script we should do something like following?
- Get the release artifact of arm also https://mirror.openshift.com/pub/openshift-v4/arm64/clients/ocp/latest-4.11/release.txt using the `arm64` location
- Get the KAO/KCMO image digest on arm64 using filter-by-os
- Use that arm64 location and digest to create new release of arm64 using amd64 machine and upload to quay?

Reply:
>There is a discrepancy between the container images which are rebuilt once per openshift release, and the release images which are rebuilt once per snc.sh run.

@cfergeau which kind of discrepancy can happen in case release image created during each snc.sh run, I hope even we run it multiple time for a specific version of Openshift it should ended up same artifacts (if not then there is some logical wrong in our approach). I do understand your point of having it part of internal script and create the release artifacts as soon as we have patched images for KAO/KCMO but that shouldn't because of any discrepancy but convenient to use with snc script and avoid pushing same content again and again (even it push just say already existing same layer).

Reply:
Fwiw, we can probably live with that for now, but just pointing out a potential simplification for this code.